Refactoring tests for corpus evaluation

rmitsch opened the pull request:
<!--- Provide a general summary of your changes in the title. -->
## Goals
- Remove repetitive code running `evaluate_corpus()`  tests.
- Central data location.
- Central minimum performance table.
- Allow for separate thresholds for different model sizes.

## Description
<!--- Use this section to describe your changes. If your changes required
testing, include information about the testing environment and the tests you
ran. If your test fixes a bug reported in an issue, don't forget to include the
issue number. If your PR is still a work in progress, that's totally fine – just
include a note to let us know. -->
- Moved test files from `tests/lang/[LANG]` into `tests/data/test_files`.
- Introduced `tests/data/performance_thresholds.csv`, which holds performance thresholds for all tested metrics in tests using `evaluate_corpus()`. Thresholds can be defined w.r.t. the used test file, model size, language and metric. See exemplary table below.
- Centralized all calls of `evaluate_corpus()` into a single, dynamically parametrized test `tests.test_corpus.test_corpus()`.
- Removed test files empty after the centralization of `evaluate_corpus()` tests.

Excerpt from `performance_thresholds.csv`:
| language | model_size | component     | filename          | metric    | accuracy |
|----------|------------|---------------|-------------------|-----------|----------|
| ca       | *          | parser        | dev001_1.json     | dep_uas   | 0.91     |   
| ca       | *          | parser        | dev001_1.json     | dep_las   | 0.88     |   
| ca       | *          | lemmatizer    | dev001_1.json     | lemma_acc | 0.95     |   
| ca       | *          | morphologizer | dev001_1.json     | pos_acc   | 0.95     |   
| ca       | *          | morphologizer | dev001_1.json     | morph_acc | 0.95     |  
| ca       | *          | ner           | dev001_1.json     | ents_f    | 0.65     |   
| da       | *          | morphologizer | ddt.dev01_10.json | pos_acc   | 0.93     | 
| da       | *          | morphologizer | ddt.dev01_10.json | morph_acc | 0.93     |

`*` indicates that the corresponding threshold applies to this test regardless of `model_size`. This can be superseded by entries with the exact model size (`sm`, `md`, `lg`, `trf`), if desired.

### Types of change
<!-- What type of change does your PR cover? Is it a bug fix, an enhancement
or new feature, or a change to the documentation? -->
Refactoring of existing code.

## Checklist
<!--- Before you submit the PR, go over this checklist and make sure you can
tick off all the boxes. [] -> [x] -->
- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

adrianeboyd commented on tests/data/performance_thresholds.csv:
```diff
@@ -0,0 +1,90 @@
+es,*,parser,es_ancora-ud-dev001_10.json,dep_uas,0.84
+es,*,parser,es_ancora-ud-dev001_10.json,dep_las,0.8
+fi,*,lemmatizer,fi_tdt-ud-dev001_1.json,lemma_acc,0.79
+fi,*,lemmatizer,fi_tdt-ud-dev001_1.json,pos_acc,0.92
+fi,*,lemmatizer,fi_tdt-ud-dev001_1.json,morph_acc,0.86
```
```suggestion
fi,*,morphologizer,fi_tdt-ud-dev001_1.json,pos_acc,0.92
fi,*,morphologizer,fi_tdt-ud-dev001_1.json,morph_acc,0.86
```

adrianeboyd commented:
It's possible I missed the details, but does this new design take the options like `--has-tagger`/`OPT_TAGGER` into account? The idea behind this is that we specify the expected configuration details externally to double-check that the pipeline isn't misconfigured, and with these options you can also just run the tagger tests and skip the other tests if desired.

rmitsch commented:
Good point! I added a check for those options now, so that the `evaluate_corpus()` tests are only executed if the corresponding options are set via CLI.

A note: "segmenter"/`--has-segmenter` wasn't included in `conftest.py` - some tests/thresholds (JA, ZH) for segmenters are defined in `performance_thresholds.csv` though. I added `OPT_SEGMENTER = "--has-segmenter"` and extended `OPT_MAPPING` & updated the readme.

svlandeg commented on tests/lang/de/test_tagger.py:
```diff
@@ -23,15 +24,6 @@ def test_de_tagger_example(NLP):
 
-# This threshold is artificially low due to problems with spacy 2.1. (#3830)
-@pytest.mark.parametrize(
-    "test_file,accuracy_threshold", [("de_pud-ud-test.stts.json", 0.94)]
```
@adrianeboyd : while we're at it, should we increase this threshold again?

adrianeboyd replied:
No, let's not modify any thresholds in this PR so that we can test v3.3.0 again with the exact same setup. Adjusting the thresholds can come later, especially for model size-specific ones.

svlandeg replied:
That makes sense, I just wanted to flag this as the "TODO" part was being removed ;-)

adrianeboyd replied:
(I'm not even sure the comment is still accurate?)

svlandeg commented on tests/test_corpus.py:
```diff
@@ -0,0 +1,19 @@
+def test_corpus(NLP, lang, corpus_evaluation_config):
+    """
+    Evaluates corpus w.r.t. the specified model and language. Information on the files to use for the individual tests,
+    metrics and and performance thresholds are extracted from performance_thresholds.csv.
```
```suggestion
    metrics and performance thresholds are extracted from performance_thresholds.csv.
```

adrianeboyd commented:
There's an error related to the csv path when you're not running it from that exact directory:

```none
_______________________ ERROR collecting test_corpus.py ________________________
../.env/lib/python3.8/site-packages/pluggy/_hooks.py:265: in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
../.env/lib/python3.8/site-packages/pluggy/_manager.py:80: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
../.env/lib/python3.8/site-packages/_pytest/python.py:249: in pytest_pycollect_makeitem
    res = list(collector._genfunctions(name, obj))
../.env/lib/python3.8/site-packages/_pytest/python.py:470: in _genfunctions
    self.ihook.pytest_generate_tests.call_extra(methods, dict(metafunc=metafunc))
../.env/lib/python3.8/site-packages/pluggy/_hooks.py:292: in call_extra
    return self(**kwargs)
../.env/lib/python3.8/site-packages/pluggy/_hooks.py:265: in __call__
    return self._hookexec(self.name, self.get_hookimpls(), kwargs, firstresult)
../.env/lib/python3.8/site-packages/pluggy/_manager.py:80: in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
tests/conftest.py:115: in pytest_generate_tests
    with open(os.path.join("data", "performance_thresholds.csv")) as threshold_file:
E   FileNotFoundError: [Errno 2] No such file or directory: 'data/performance_thresholds.csv'
=========================== short test summary info ============================
ERROR tests/test_corpus.py - FileNotFoundError: [Errno 2] No such file or dir...
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.31s ===============================
```

I think you need to use a path that's relative to the test module (this may not be the best/only way):

https://github.com/explosion/spacy-models/blob/3d347dfd1755a004cf9b686edbffbfbec51515d8/tests/lang/de/test_morphologizer.py#L6

rmitsch commented:
Ah, that slipped through the cracks. Fixed.

adrianeboyd commented:
I think we can go ahead and merge as it looks equivalent to the previous tests, and then in future PRs we can work on more accurate thresholds for the full range of models and maybe some test corpora for languages that are currently missing.
